Avoid make subqueries from escape semicolons (issue #128 and #149)

stillalivx opened the pull request:
Hi! I found a solution to the issue #128. 

I just replaced the statement ```split(';')``` to this regex ```/(?<!\\);/```. That avoid to the function to split a query if the semicolon was escaped. I also make a pull request to Dex to make the escape of semicolon by "default".

eveningkid commented:
Hey,

Thanks for helping on this! This was actually quite an odd one.

I tried your solution with the following fake query and here's what I got:

```
> let q = 'SELECT * FROM friends WHERE message = "hello;world"; UPDATE friends'
'SELECT * FROM friends WHERE message = "hello;world"; UPDATE friends'
> q.split(/(?<!\\);/)
[
  'SELECT * FROM friends WHERE message = "hello',
  'world"',
  ' UPDATE friends'
]
```

Do you know how it could also work for this use case?

stillalivx commented:
You have to escape the semicolon before make the query. 
```
> let q = 'SELECT * FROM friends WHERE message = "hello\\;world"; UPDATE friends'
'SELECT * FROM friends WHERE message = "hello\\;world"; UPDATE friends'
> q.split(/(?<!\\);/)
[
  'SELECT * FROM friends WHERE message = "hello',
  'world"',
  ' UPDATE friends'
]
```
Here we have a issue with Dex, 'cause Dex escape de '\' character by defualt. I made a little change in Dex repo to escape the semicolons, and that make works fine! Here a let you the commit (https://github.com/denjucks/dex/pull/11). I don't know if they gonna merge it, I saw that the repo is not very active. 

In a project, I refer to denodb export file to my fork of Dex to make that work. I haven't had problems there. (https://github.com/stillalivx/denodb/commit/31636bed07cd623760f82683ddf4517afd65e5c6)

eveningkid commented:
Thanks for the quick reply.

I just tried with the `;` being escaped and I get the following (which seems to work):

```
> q = 'SELECT * FROM friends WHERE message = "hello\\;world"; UPDATE friends'
'SELECT * FROM friends WHERE message = "hello\\;world"; UPDATE friends'
> q.split(/(?<!\\);/)
[
  'SELECT * FROM friends WHERE message = "hello\\;world"',
  ' UPDATE friends'
]
```

Now I have a question: if the DB engine gets `message = "hello\\;world"`, will it remove the two backslashes?

If so, your solution is safe and we should go with our own fork/replication of Dex since it's been inactive for +7 months..!
And only then, we'll just need to update the dex dependency first to the good version and merge this PR.

I really hope it works! :)

stillalivx commented:
Hi! I updated the regex. With this one we don't need to escape the semicolon. Actually, is not neccesary to make the changes on Dex's repo.
I changed the regex because on SQLite engine can't escape se semicolon like MariaDB (or similar).

Let you a little code and its expected result:
```
const query = "SELECT * FROM tableName WHERE 'Value;Other;Value';UPDATE tableName WHERE 'Value;Other;Value'";
> "SELECT * FROM tableName WHERE 'Value;Other;Value';UPDATE tableName WHERE 'Value;Other;Value'"
query.split(/;(?=(?:[^'"]|'[^']*'|"[^"]*")*$)/);
> [
  "SELECT * FROM tableName WHERE 'Value;Other;Value'",
  "UPDATE tableName WHERE 'Value;Other;Value'",
]
```
PS: I don't know why the test failed. Very odd. 🤔
